### PR TITLE
Refactor and unit-test FreeRTOS_IP.c

### PR DESCRIPTION
**Pull request by AniruddhaKanhere**
Description
-----------
This PR splits FreeRTOS_IP.c into multiple files to increase modularity and to increase ease of testing. Namely:
- FreeRTOS_IP.c - Core functionality and all `FreeRTOS_xxx` APIs.
- FreeRTOS_ICMP.c - Processing of ICMP packets.
- FreeRTOS_IP_Utils.c - Utility functions for IP. Such as `FreeRTOS_max_uint32` etc.
- FreeRTOS_IP_Timers.c - FreeRTOS_IP used to process timer events for ARP/DHCP/IP/TCP. This module now handles the timer related work.

In addition, this PR adds 100% unit-test coverage for FreeRTOS_IP.c and FreeRTOS_ICMP.c.
There will be a PR later which will add coverage for the remaining two files.

Test Steps
-----------

Related Issue
-----------

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: include/FreeRTOS_IP_Private.h [hunk 1]
```diff
@@ -916,21 +922,6 @@
 
     #endif /* ipconfigSUPPORT_SELECT_FUNCTION */
 
-    #if ( ipconfigSUPPORT_SELECT_FUNCTION == 1 ) || ( ipconfigUSE_TCP == 1 ) || ( ipconfigDNS_USE_CALLBACKS == 1 )
-        extern ipDECL_CAST_PTR_FUNC_FOR_TYPE( ListItem_t );
-        extern ipDECL_CAST_CONST_PTR_FUNC_FOR_TYPE( ListItem_t );
-    #endif
-
-    extern ipDECL_CAST_PTR_FUNC_FOR_TYPE( NetworkBufferDescriptor_t );
```
Comment (htibosch) [inline include/FreeRTOS_IP_Private.h:924]: Now when I compile Sockets.c, it is missing the above CAST functions. Was that a mistake to (re)move them?

File: FreeRTOS_IP.c [hunk 2]
```diff
@@ -41,6 +41,8 @@
 
 /* FreeRTOS+TCP includes. */
 #include "FreeRTOS_IP.h"
+#include "FreeRTOS_ICMP.h"
+#include "FreeRTOS_IP_Timers.h"
 #include "FreeRTOS_Sockets.h"
```
Comment (htibosch) [inline FreeRTOS_IP.c:46]: Missing: `#include "FreeRTOS_IP_Utils.h"` 
If not, `prvProcessNetworkDownEvent()` and other identifiers would be "undeclared".

File: FreeRTOS_IP_Utils.c [hunk 3]
```diff
@@ -0,0 +1,1208 @@
...
+
+/* FreeRTOS includes. */
+#include "FreeRTOS.h"
+#include "task.h"
+#include "queue.h"
+#include "semphr.h"
+
+/* FreeRTOS+TCP includes. */
+#include "FreeRTOS_IP.h"
+#include "FreeRTOS_Sockets.h"
```
Comment (htibosch) [inline FreeRTOS_IP_Utils.c:44]: It will also need to include `FreeRTOS_IP_Utils.h` and `FreeRTOS_IP_Timers.h`

File: FreeRTOS_IP_Utils.c [hunk 4]
```diff
@@ -0,0 +1,1208 @@
...
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ *
+ * http://aws.amazon.com/freertos
+ * http://www.FreeRTOS.org
+ */
+
+/**
+ * @file FreeRTOS_IP.c
+ * @brief Implements the basic functionality for the FreeRTOS+TCP network stack.
+ */
```
Comment (htibosch) [inline FreeRTOS_IP_Utils.c:29]: Change name to FreeRTOS_IP_Utils.c ?

File: FreeRTOS_IP_Timers.c [hunk 5]
```diff
@@ -0,0 +1,468 @@
...
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ *
+ * http://aws.amazon.com/freertos
+ * http://www.FreeRTOS.org
+ */
+
+/**
+ * @file FreeRTOS_ICMP.c
+ * @brief Implements the Internet Control Message Protocol for the FreeRTOS+TCP network stack.
+ */
```
Comment (htibosch) [inline FreeRTOS_IP_Timers.c:29]: Change name to FreeRTOS_IP_Timers.c ?

**Review by htibosch — COMMENTED** [review]
Hi Aniruddha, thanks for this PR so far.

I tried to run the MISRA checks and found it difficult to compile. I had about 16 corrections, all described here.

Is it really important or useful to add new header files?

And if so, can you please include these files automatically: `FreeRTOS_IP_Timers`, `FreeRTOS_IP_Utils.h`, and `FreeRTOS_ICMP.h`?

- Many applications will now have to include `FreeRTOS_ICMP.h`, which breaks their code.
- Network interfaces call a function `vPrintResourceStats()`, which is now defined in `FreeRTOS_IP_Utils.h`. Also `prvProcessNetworkDownEvent()` got "undefined".

I would like to propose to keep using the new headers files, but without breaking any existing project.

Thanks,

**Review by alfred2g — APPROVED** [review]

**Comment by AniruddhaKanhere** [comment]
@htibosch, this PR is a part of two PRs adding separate unit-tests.
The breaking build issues will be addressed in the next PR.

**Review by alfred2g — APPROVED** [review]

**Review by cobusve — APPROVED** [review]

**Pull request merged**